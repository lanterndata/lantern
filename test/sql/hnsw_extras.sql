------------------------------------------------------------------------------
-- Test Functions exported from lantern_extras extension
------------------------------------------------------------------------------

\ir utils/sift1k_array.sql

\set ON_ERROR_STOP off
CREATE EXTENSION lantern_extras;
-- Validate error on invalid params
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'invalid_metric',  3, 10, 10, 10);
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'l2sq',  3, -1, 10, 10);
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'l2sq',  3, 10, -2, 10);
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'l2sq',  3, 10, 10, -1);

-- Validate error on empty table
CREATE TABLE empty (v REAL[]);
SELECT lantern_create_external_index('v', 'empty');
\set ON_ERROR_STOP on

-- Create with defaults
SELECT lantern_create_external_index('v', 'sift_base1k');
SELECT _lantern_internal.validate_index('sift_base1k_v_idx', false);

SELECT v AS v777 FROM sift_base1k WHERE id = 777 \gset
-- Validate that using corresponding operator triggers index scan
SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <-> :'v777' LIMIT 10;

SET lantern.pgvector_compat=FALSE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <?> :'v777' LIMIT 10;
SET lantern.pgvector_compat=TRUE;
DROP INDEX sift_base1k_v_idx;

-- Create with params
SELECT lantern_create_external_index('v', 'sift_base1k', 'public', 'cos', 128, 10, 10, 10, false, 'hnsw_cos_index');
SELECT _lantern_internal.validate_index('hnsw_cos_index', false);
SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;

SET lantern.pgvector_compat=FALSE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <?> :'v777' LIMIT 10;
SET lantern.pgvector_compat=TRUE;

-- -- Reindex external index
SELECT lantern_reindex_external_index('hnsw_cos_index');
SELECT _lantern_internal.validate_index('hnsw_cos_index', false);

-- Validate that using corresponding operator triggers index scan
SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;

-- Create PQ Index
SET client_min_messages=ERROR;
DROP INDEX hnsw_cos_index;
-- Verify error that codebook does not exist
\set ON_ERROR_STOP off
SELECT lantern_create_external_index('v', 'sift_base1k', 'public', 'cos', 128, 10, 10, 10, true, 'hnsw_cos_index_pq');
\set ON_ERROR_STOP on
SELECT quantize_table('sift_base1k'::regclass, 'v', 10, 32, 'cos');
SELECT lantern_create_external_index('v', 'sift_base1k', 'public', 'cos', 128, 10, 10, 10, true, 'hnsw_cos_index_pq');
SELECT _lantern_internal.validate_index('hnsw_cos_index_pq', false);
SELECT lantern_reindex_external_index('hnsw_cos_index_pq');
SELECT _lantern_internal.validate_index('hnsw_cos_index_pq', false);
SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;

SELECT drop_quantization('sift_base1k'::regclass, 'v');
DROP INDEX hnsw_cos_index_pq;

-- Create using CREATE INDEX syntax with defaults
CREATE INDEX hnsw_l2_index ON sift_base1k USING lantern_hnsw(v) WITH (external=true);
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <-> :'v777' LIMIT 10;
SELECT _lantern_internal.validate_index('hnsw_l2_index', false);
DROP INDEX hnsw_l2_index;
-- Create using CREATE INDEX syntax with params
CREATE INDEX hnsw_cos_index ON sift_base1k USING lantern_hnsw(v dist_cos_ops) WITH (m=12, ef=128, ef_construction=32, external=true);
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;
SELECT _lantern_internal.validate_index('hnsw_cos_index', false);
DROP INDEX hnsw_cos_index;
-- Create using CREATE INDEX syntax with pq (should error)
\set ON_ERROR_STOP off
CREATE INDEX hnsw_cos_index ON sift_base1k USING lantern_hnsw(v dist_cos_ops) WITH (m=12, ef=128, ef_construction=32, external=true, pq=true);
\set ON_ERROR_STOP on
SELECT quantize_table('sift_base1k'::regclass, 'v', 10, 32, 'cos');
CREATE INDEX hnsw_cos_index_pq ON sift_base1k USING lantern_hnsw(v dist_cos_ops) WITH (m=12, ef=128, ef_construction=32, external=true, pq=true);
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;
SELECT _lantern_internal.validate_index('hnsw_cos_index_pq', false);
DROP INDEX hnsw_cos_index_pq;

