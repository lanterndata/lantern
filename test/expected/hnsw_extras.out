------------------------------------------------------------------------------
-- Test Functions exported from lantern_extras extension
------------------------------------------------------------------------------
\ir utils/sift1k_array.sql
CREATE TABLE IF NOT EXISTS sift_base1k (
    id SERIAL,
    v REAL[]
);
COPY sift_base1k (v) FROM '/tmp/lantern/vector_datasets/sift_base1k_arrays.csv' WITH csv;
\set ON_ERROR_STOP off
CREATE EXTENSION lantern_extras;
-- Validate error on invalid params
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'invalid_metric',  3, 10, 10, 10);
ERROR:  Invalid metric invalid_metric
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'l2sq',  3, -1, 10, 10);
ERROR:  m should be in range [2, 128]
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'l2sq',  3, 10, -2, 10);
ERROR:  ef_construction should be in range [1, 400]
SELECT lantern_create_external_index('v','sift_base1k', 'public', 'l2sq',  3, 10, 10, -1);
ERROR:  ef should be in range [1, 400]
-- Validate error on empty table
CREATE TABLE empty (v REAL[]);
SELECT lantern_create_external_index('v', 'empty');
ERROR:  Cannot create an external index on empty table
\set ON_ERROR_STOP on
-- Create with defaults
SELECT lantern_create_external_index('v', 'sift_base1k');
 lantern_create_external_index 
-------------------------------
 
(1 row)

SELECT _lantern_internal.validate_index('sift_base1k_v_idx', false);
INFO:  validate_index() start for sift_base1k_v_idx
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

SELECT v AS v777 FROM sift_base1k WHERE id = 777 \gset
-- Validate that using corresponding operator triggers index scan
SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <-> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using sift_base1k_v_idx on sift_base1k
         Order By: (v <-> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SET lantern.pgvector_compat=FALSE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <?> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using sift_base1k_v_idx on sift_base1k
         Order By: (v <?> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SET lantern.pgvector_compat=TRUE;
DROP INDEX sift_base1k_v_idx;
-- Create with params
SELECT lantern_create_external_index('v', 'sift_base1k', 'public', 'cos', 128, 10, 10, 10, false, 'hnsw_cos_index');
 lantern_create_external_index 
-------------------------------
 
(1 row)

SELECT _lantern_internal.validate_index('hnsw_cos_index', false);
INFO:  validate_index() start for hnsw_cos_index
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using hnsw_cos_index on sift_base1k
         Order By: (v <=> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SET lantern.pgvector_compat=FALSE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <?> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using hnsw_cos_index on sift_base1k
         Order By: (v <?> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SET lantern.pgvector_compat=TRUE;
-- -- Reindex external index
SELECT lantern_reindex_external_index('hnsw_cos_index');
 lantern_reindex_external_index 
--------------------------------
 
(1 row)

SELECT _lantern_internal.validate_index('hnsw_cos_index', false);
INFO:  validate_index() start for hnsw_cos_index
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

-- Validate that using corresponding operator triggers index scan
SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using hnsw_cos_index on sift_base1k
         Order By: (v <=> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

-- Create PQ Index
SET client_min_messages=ERROR;
DROP INDEX hnsw_cos_index;
-- Verify error that codebook does not exist
\set ON_ERROR_STOP off
SELECT lantern_create_external_index('v', 'sift_base1k', 'public', 'cos', 128, 10, 10, 10, true, 'hnsw_cos_index_pq');
ERROR:  Codebook table "_lantern_internal"."pq_sift_base1k_v" does not exist
\set ON_ERROR_STOP on
SELECT quantize_table('sift_base1k'::regclass, 'v', 10, 32, 'cos');
INFO:  Table scanned. Dataset size 1000
INFO:  Starting k-means over dataset with (subvectors=32, clusters=10)
INFO:  Codebooks created
INFO:  Compressing vectors...
 quantize_table 
----------------
 
(1 row)

SELECT lantern_create_external_index('v', 'sift_base1k', 'public', 'cos', 128, 10, 10, 10, true, 'hnsw_cos_index_pq');
 lantern_create_external_index 
-------------------------------
 
(1 row)

SELECT _lantern_internal.validate_index('hnsw_cos_index_pq', false);
INFO:  validate_index() start for hnsw_cos_index_pq
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

SELECT lantern_reindex_external_index('hnsw_cos_index_pq');
 lantern_reindex_external_index 
--------------------------------
 
(1 row)

SELECT _lantern_internal.validate_index('hnsw_cos_index_pq', false);
INFO:  validate_index() start for hnsw_cos_index_pq
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

SET lantern.pgvector_compat=TRUE;
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using hnsw_cos_index_pq on sift_base1k
         Order By: (v <=> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SELECT drop_quantization('sift_base1k'::regclass, 'v');
 drop_quantization 
-------------------
 
(1 row)

DROP INDEX hnsw_cos_index_pq;
-- Create using CREATE INDEX syntax with defaults
CREATE INDEX hnsw_l2_index ON sift_base1k USING lantern_hnsw(v) WITH (external=true);
INFO:  done init usearch index
INFO:  done loading usearch index
INFO:  done saving 1000 vectors
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <-> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using hnsw_l2_index on sift_base1k
         Order By: (v <-> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SELECT _lantern_internal.validate_index('hnsw_l2_index', false);
INFO:  validate_index() start for hnsw_l2_index
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

DROP INDEX hnsw_l2_index;
-- Create using CREATE INDEX syntax with params
CREATE INDEX hnsw_cos_index ON sift_base1k USING lantern_hnsw(v dist_cos_ops) WITH (m=12, ef=128, ef_construction=32, external=true);
INFO:  done init usearch index
INFO:  done loading usearch index
INFO:  done saving 1000 vectors
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using hnsw_cos_index on sift_base1k
         Order By: (v <=> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SELECT _lantern_internal.validate_index('hnsw_cos_index', false);
INFO:  validate_index() start for hnsw_cos_index
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

DROP INDEX hnsw_cos_index;
-- Create using CREATE INDEX syntax with pq (should error)
\set ON_ERROR_STOP off
CREATE INDEX hnsw_cos_index ON sift_base1k USING lantern_hnsw(v dist_cos_ops) WITH (m=12, ef=128, ef_construction=32, external=true, pq=true);
ERROR:  PQ-codebook for relation "sift_base1k" not found
\set ON_ERROR_STOP on
SELECT quantize_table('sift_base1k'::regclass, 'v', 10, 32, 'cos');
INFO:  Table scanned. Dataset size 1000
INFO:  Starting k-means over dataset with (subvectors=32, clusters=10)
INFO:  Codebooks created
INFO:  Compressing vectors...
 quantize_table 
----------------
 
(1 row)

CREATE INDEX hnsw_cos_index_pq ON sift_base1k USING lantern_hnsw(v dist_cos_ops) WITH (m=12, ef=128, ef_construction=32, external=true, pq=true);
INFO:  done init usearch index
INFO:  done loading usearch index
INFO:  done saving 1000 vectors
EXPLAIN (COSTS FALSE) SELECT id FROM sift_base1k order by v <=> :'v777' LIMIT 10;
                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   ->  Index Scan using hnsw_cos_index_pq on sift_base1k
         Order By: (v <=> '{97,67,0,0,0,0,0,14,49,107,23,0,0,0,5,24,4,25,48,5,0,1,8,3,0,5,17,3,1,1,3,3,126,126,0,0,0,0,0,27,49,126,49,8,1,4,11,14,0,6,37,39,10,22,25,0,0,0,12,27,7,23,35,3,126,9,1,0,0,0,19,126,28,11,8,7,1,39,126,126,0,1,28,27,3,126,126,0,1,3,7,9,0,52,126,5,13,5,8,0,0,0,33,72,78,19,18,3,0,3,21,126,42,13,64,83,1,9,8,23,1,4,22,68,3,1,4,0}'::real[])
(3 rows)

SELECT _lantern_internal.validate_index('hnsw_cos_index_pq', false);
INFO:  validate_index() start for hnsw_cos_index_pq
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

DROP INDEX hnsw_cos_index_pq;
